accessors/spanner/spanner_accessor.go

```diff
@@ -201,3 +233,211 @@ func (sp *SpannerAccessorImpl) CreateChangeStream(ctx context.Context, adminClie
+		migrationData := metrics.GetMigrationData(conv, driver, constants.SchemaConv)
+		serializedMigrationData, _ := proto.Marshal(migrationData)
+		migrationMetadataValue := base64.StdEncoding.EncodeToString(serializedMigrationData)
+		ctx = metadata.AppendToOutgoingContext(ctx, constants.MigrationMetadataKey, migrationMetadataValue)
+	}
+	if dbExists {
+		if conv.SpDialect != constants.DIALECT_POSTGRESQL && migrationType == constants.DATAFLOW_MIGRATION {
+			return fmt.Errorf("spanner migration tool does not support minimal downtime schema/schema-and-data migrations to an existing database")
+		}
+		err := sp.UpdateDatabase(ctx, adminClient, dbURI, conv, driver)
+		if err != nil {
+			return fmt.Errorf("can't update database schema: %v", err)
+		}
+	} else {
+		err := sp.CreateDatabase(ctx, adminClient, dbURI, conv, driver, migrationType)
+		if err != nil {
+			return fmt.Errorf("can't create database: %v", err)
+		}
+	}
+	return nil
+}
+
+// VerifyDb checks whether the db exists and if it does, verifies if the schema is what we currently support.
+func (sp *SpannerAccessorImpl) VerifyDb(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string) (dbExists bool, err error) {
+	dbExists, err = sp.CheckExistingDb(ctx, adminClient, dbURI)
+	if err != nil {
+		return dbExists, err
+	}
+	if dbExists {
+		err = sp.ValidateDDL(ctx, adminClient, dbURI)
+	}
+	return dbExists, err
+}
+
+// ValidateDDL verifies if an existing DB's ddl follows what is supported by Spanner migration tool. Currently,
+// we only support empty schema when db already exists.
+func (sp *SpannerAccessorImpl) ValidateDDL(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string) error {
+	dbDdl, err := adminClient.GetDatabaseDdl(ctx, &adminpb.GetDatabaseDdlRequest{Database: dbURI})
+	if err != nil {
+		return fmt.Errorf("can't fetch database ddl: %v", err)
+	}
+	if len(dbDdl.Statements) != 0 {
+		return fmt.Errorf("spanner migration tool supports writing to existing databases only if they have an empty schema")
+	}
+	return nil
+}
+
+
+// UpdateDDLForeignKeys updates the Spanner database with foreign key
+// constraints using ALTER TABLE statements.
+func (sp *SpannerAccessorImpl) UpdateDDLForeignKeys(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string, conv *internal.Conv, driver string, migrationType string) {
+
+	if conv.SpDialect != constants.DIALECT_POSTGRESQL && migrationType == constants.DATAFLOW_MIGRATION {
+		//foreign keys were applied as part of CreateDatabase
+		return
+	}
+
+	// The schema we send to Spanner excludes comments (since Cloud
+	// Spanner DDL doesn't accept them), and protects table and col names
+	// using backticks (to avoid any issues with Spanner reserved words).
+	fkStmts := conv.SpSchema.GetDDL(ddl.Config{Comments: false, ProtectIds: true, Tables: false, ForeignKeys: true, SpDialect: conv.SpDialect, Source: driver})
+	if len(fkStmts) == 0 {
+		return
+	}
+	if len(fkStmts) > 50 {
+		fmt.Println(`
```

Comment:
This is also an opportunity to replace these print statements with zap logger. It'd be an easy win to replace these print statements as we go.